Remove dead code in RSA

valeriosetti opened the pull request:
## Description

Resolve #110

## PR checklist

- [x] **changelog** [not required](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/559#discussion_r2489840193)
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: no change there
- [x] **mbedtls 3.6 PR** not required because: no backport
- **tests**  provided: they have been updated as requested on the related issue

valeriosetti committed:
drivers: builtin: rsa: remove mbedtls_rsa_complete()

This functionality is no more required now that RSA is a builtin module.
All the parameters are always provided.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: builtin: rsa: remove usage of mbedtls_rsa_complete()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: builtin: rsa: check private key in mbedtls_rsa_parse_key

Since we're parsing a private key where we assume all the necessary
parameters have been provided (because mbedtls_rsa_complete() was removed),
it's better to call mbedtls_rsa_check_privkey() instead of
mbedtls_rsa_check_pubkey().

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkcs1|rsa: replace usage of mbedtls_rsa_complete() in test functions

Replace mbedtls_rsa_complete() with mbedtls_rsa_deduce_private_exponent()
and mbedtls_rsa_deduce_crt()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: remove test cases with incomplete parameters

mbedtls_rsa_complete() has been removed since now all input parameters
must be provided when importing an RSA key.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: fix return value for private key parsing test case

After the removal of mbedtls_rsa_complete() from mbedtls_rsa_parse_key()
now the latter returns MBEDTLS_ERR_RSA_KEY_CHECK_FAILED instead of
MBEDTLS_ERR_RSA_BAD_INPUT_DATA in case of invalid parameters (including
even modulus).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: remove remaining occurrences of mbedtls_rsa_complete()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg commented:
> * **changelog** maybe required: there is 1 or 2 test cases where the expected error code changes. I guess this needs a changelog, right?

I'm not sure. We don't [promise](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#backwards-compatibility-for-application-code) stability for which error code is return when more than one applies, so hopefully applications should not be relying on that, and should not need to be informed of the change by a ChangeLog entry.

OTOH, if we think it is likely that a reasonable application was checking for a specific error code, then IMO the answer is not really to add a ChangeLog entry but to re-consider whether that change really is suitable for a .x release (ie, not a major version).

Probably depends on the specifics, I'll probably be able to have a more precise opinion after actually reviewing the code :)

mpg commented on drivers/builtin/src/rsa.c:
```diff
@@ -2763,6 +2764,13 @@ int mbedtls_rsa_self_test(int verbose)
     MBEDTLS_MPI_CHK(mbedtls_rsa_import(&rsa, NULL, NULL, NULL, NULL, &K));
 
+#if !defined(MBEDTLS_RSA_NO_CRT)
+    MBEDTLS_MPI_CHK(mbedtls_rsa_deduce_crt(&rsa.P, &rsa.Q, &rsa.D, &DP, &DQ, &QP));
```
Minor: why do we need intermediate variables here? Why not directly use `&rsa.DP` etc as the output parameters in this call?

mpg commented on tests/suites/test_suite_pkcs1_v15.function:
```diff
@@ -92,10 +95,18 @@ void pkcs1_rsaes_v15_decrypt(int mod, char *input_P, char *input_Q,
+    TEST_ASSERT(mbedtls_rsa_import(&ctx, &N, &P, &Q, &D, &E) == 0);
+
+#if !defined(MBEDTLS_RSA_NO_CRT)
+    TEST_ASSERT(mbedtls_rsa_deduce_crt(&P, &Q, &D, &DP, &DQ, &QP) == 0);
```
Same (and all other calls in tests, I won't comment on each one).

Pre-existing: given how many times this happens, I wonder if we should have a helper function for "populate an RSA context from a set of  test input numbers".

valeriosetti replied:
Good idea indeed. I added some commits at the end of this PR to implement this.

valeriosetti replied:
Notice: in order to reduce code duplication as much as possible and re-use it in different test suite files I added helper functions to `framework`. Unfortunately this changed this PR from a standalone one to being the last one of a series of 3. I updated the description accordingly.

gilles-peskine-arm replied:
How about merging `test_suite_pkcs1_v15.function` and `test_suite_pkcs1_v21.function` into `test_suite_rsa.function`? (We can still keep separate `.data` files.) I never understood why RSA got three test suites when generally each module gets one.

valeriosetti replied:
That's a good plan as well as it allows this PR to remove all of its dependencies

mpg replied:
Yes, I think that's the simplest thing to do, keeps everything in a single repo.

valeriosetti replied:
> (We can still keep separate .data files.)

I suspect they need to be renamed though to something like `test_suite_rsa.pkc1_v15.data`, otherwise the build system will complain that `test_suite_pkcs1_v15.function` does not exist. If this is the case then I think I'll still need to open a PR in `mbedtls` because I checked `anaylyze_outcomes.py` and there are exceptions for these pkcs1 files.

gilles-peskine-arm replied:
Right, they would need to be renamed.

Fortunately I don't think mbedtls's `analyze_outcomes.py` needs changing: it just skips the whole suites `'rsa', 'pkcs1_v15', 'pkcs1_v21'` when doing RSA testing. And ignoring a non-existing suite is not flagged as an error (what's flagged as an error is if something is ignored but the driver test passes anyway).

valeriosetti replied:
Yeah ignoring a suite that does not exist is not an error, but I suspect that when it will come to test `DriverVSReference_rsa` it won't skip disparities between reference and driver for accelerated RSA and that will be marked as error. Anyway let's see what the CI thinks about this :)

mpg commented on tests/suites/test_suite_rsa.data:
```diff
@@ -592,7 +592,7 @@ RSA parse private key - correct format, modulus too small (127 bit)
 
 RSA parse private key - correct format, modulus even
-rsa_parse_pkcs1_key:0:"3063020100021100cc8ab070369ede72920e5a51523c857002030100010211009a6318982a7231de1894c54aa4909201020900f3058fd8dc484d61020900d7770dbd8b78a2110209009471f14c26428401020813425f060c4b72210208052b93d01747a87c":MBEDTLS_ERR_RSA_BAD_INPUT_DATA
+rsa_parse_pkcs1_key:0:"3063020100021100cc8ab070369ede72920e5a51523c857002030100010211009a6318982a7231de1894c54aa4909201020900f3058fd8dc484d61020900d7770dbd8b78a2110209009471f14c26428401020813425f060c4b72210208052b93d01747a87c":MBEDTLS_ERR_RSA_KEY_CHECK_FAILED
```
Note: I think that's an acceptable change of error code, which does not need a changelog entry.

mpg reviewed (commented):
What's there is looking good so far (except for the issue with testing overly large keys), just a few minor points.

However I don't think this PR in its current state completely resolves #110. For example unless I'm mistaken `mbedtls_rsa_deduce_primes()` should be dead code now that `mbedtls_rsa_complete()` is removed. (I just had a look, and that's the only one I could find. Since it's now fully internal, its removal should be transparent.)

Also, we could go further and merge back `rsa_alt_helpers.c` into `rsa.c`, making those functions `static`. However, I think that would be scope creep here. Let's keep this PR focused on removing dead code, and I'll file another issue for other simplifications we can make in RSA code.

valeriosetti committed:
drivers|tests: rsa: check bit-size on key pair import

This fix is required to make the CI happy on test
'PSA import RSA key pair: maximum size exceeded' from
'test_suite_psa_crypto.data'. As explained in the comments added in this
commit, issue tf-psa-crypto#562 has been created to track this testing
weakness and to solve it in the future.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: rsa: remove intermediate mpi values in mbedtls_rsa_self_test()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa|pkcs1: use new helper function to fill RSA contexts

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: fix public exponent format

This value must be parsable from unhexify so "3" -> "03"

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers|tests: rsa: remove mbedtls_rsa_deduce_primes()

The function is no more used anywhere in driver code so it's dead code.
This commit removes function definition, implementation and related
test code.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
> Also, we could go further and merge back `rsa_alt_helpers.c` into `rsa.c`, making those functions `static`. However, I think that would be scope creep here. Let's keep this PR focused on removing dead code, and I'll file another issue for other simplifications we can make in RSA code.

Currently `rsa.c` is already a quite long source file (close to 3k lines). Personally I would prefer to keep helpers in a separate file and actually create a new one for signature/encryption, but I understand this is a matter of preference.

valeriosetti committed:
tests: rsa: move test_suite_pkcs1_v15 test functions to test_suite_rsa

File test_suite_pkcs1_v15.function is then removed as it's empty.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: move test_suite_pkcs1_v21 test functions to test_suite_rsa

File test_suite_pkcs1_v21.function is then removed as it's empty.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: rename pkcs1_[v15|v21] data files

Since test functions now live in "test_suite_rsa", the data file names must
start with "test_suite_rsa". The "pkcs1_[v15|v21]" is kept after the 1st
dot.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: move RSA helper function to test_suite_rsa

Initially this was placed in the framework repo, but this turned out to
be a bad idea. This commit brings the implementation back to test_suite_rsa
which is the only place where that helper function is used.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: re-add mbedtls_rsa_free() in pkcs1_rsassa_pss_sign()

This was erroneously removed in previous commits.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg commented on drivers/builtin/src/rsa.c:
```diff
@@ -51,9 +51,7 @@
- * The value zero is:
- * - never a valid value for an RSA parameter
- * - interpreted as "omitted, please reconstruct" by mbedtls_rsa_complete().
+ * The value zero is never a valid value for an RSA parameter.
```
Two lines below: remaining mention of `rsa_complete()`.

mpg commented on tests/suites/test_suite_rsa.data:
```diff
@@ -481,237 +481,75 @@ mbedtls_rsa_import:"b38ac65c8141f7f5c96e14470e851936a67bf94cc6821a39ac12c05f7c0b
 RSA Export (N,P,Q,D,E), successive
 mbedtls_rsa_export:"b38ac65c8141f7f5c96e14470e851936a67bf94cc6821a39ac12c05f7c0b06d9e6ddba2224703b02e25f31452f9c4a8417b62675fdc6df46b94813bc7b9769a892c482b830bfe0ad42e46668ace68903617faf6681f4babf1cc8e4b0420d3c7f61dc45434c6b54e2c3ee0fc07908509d79c9826e673bf8363255adb0add2401039a7bcd1b4ecf0fbe6ec8369d2da486eec59559dd1d54c9b24190965eafbdab203b35255765261cd0909acf93c3b8b8428cbb448de4715d1b813d0c9482 ...
 
-RSA Export (N,-,-,D,E)
```
I don't see a reason to remove specifically the tests about partial export. Partial import requires `rsa_complete()` which we're removing, but partial export works on its own.

Arguably, one could just remove the import and export functions and replace them with calls to `mpi_copy()` and the like (ie, `mpi_read/write_binary` for `import/export_raw`), but that would be expanding the scope of what's already a fairly large PR, so I wouldn't recommend it now (though it could be done as follow-up).

valeriosetti replied:
> I don't see a reason to remove specifically the tests about partial export. Partial import requires rsa_complete() which we're removing, but partial export works on its own.

I removed them because they were using `mbedtls_rsa_complete()` in `mbedtls_rsa_export()` to reconstruct the entire RSA context from missing parameters. `mbedtls_rsa_export[_raw]()` is still tested, but only when all the parameters are provided.

> Arguably, one could just remove the import and export functions and replace them with calls to mpi_copy() and the like (ie, mpi_read/write_binary for import/export_raw), but that would be expanding the scope of what's already a fairly large PR, so I wouldn't recommend it now (though it could be done as follow-up).

Is there an issue for this? Because I quickly tried to search for it, but I didn't find any. If not I can file the issue to track this

mpg replied:
> I removed them because they were using `mbedtls_rsa_complete()` in `mbedtls_rsa_export()` to reconstruct the entire RSA context from missing parameters.

Ah, I had missed that, sorry. Then I agree it makes complete sense to remove those test cases :)

> Is there an issue for this? Because I quickly tried to search for it, but I didn't find any. If not I can file the issue to track this

I don't think there's an issue - thanks for offering to create one. (I think you can put it in the "shared backlog" column on the EPICs board.) For context, the purpose of these functions, back when `MBEDTLS_RSA_ALT` was a thing, was to abstract away the fields of `rsa_context`, because with `MBEDTLS_RSA_ALT` we had no idea what the definition of that structure was. But now that we've removed that option, we know we're using our own definition for the structure, so we can access the fields directly.

valeriosetti replied:
Done https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/576

mpg commented on tests/suites/test_suite_rsa.function:
```diff
@@ -5,7 +5,42 @@
+#if defined(MBEDTLS_RSA_C) && defined(MBEDTLS_BIGNUM_C)
+
+static int mbedtls_rsa_test_fill_context(mbedtls_rsa_context *rsa,
+                                         const unsigned char *P_data, size_t P_len,
```
Minor: in practice I think all callers have `data_t` - if that's the case perhaps it would be a bit simpler to also have `data_t` parameters here?

valeriosetti replied:
Yes, you're absolutely right. I think I didn't do that initially because this function was placed somewhere else where `data_t` type wasn't available, but now that it lives in a `*.function` file we can make this look better.

mpg reviewed (commented):
Looks pretty good to me, just a few minor suggestions.

valeriosetti committed:
drivers: rsa: remove mention to rsa_complete from doc of asn1_get_nonzero_mpi()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: rsa: use data_t for RSA parameters in mbedtls_rsa_test_fill_context()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (approved):
LGTM, thanks!
